### PR TITLE
Packet datasource

### PR DESCRIPTION
**Pull request by stresler**
Feedback welcome. Thanks.

### CODE REVIEW

**Comment by crawford** [comment]
You also need to add the "packet" OEM to the list in coreos-cloudinit.go. I see you are calling it from the OEM cloudinit in another PR.

**Comment by crawford** [comment]
Make sure you also plumb through the appropriate flags in the main file. Otherwise, this is on the right track.

**Comment by stresler** [comment]
Totally forgot to commit a file.   One moment.

**Comment by stresler** [comment]
This probably warrants some discussion. 

We don't really use the coreos-cloudinit/network/packet.go in production. It is used on first boot to grab network data from metadata service and write it to disk. 

It seems like that would be a good use case for Ignition and that whole section of this can probably move there in short order, but for current stable this seems the right place for it.

File: datasource/metadata/packet/metadata.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+	userdataUrl    = "userdata"
+	metadataPath   = "metadata"
+)
+
+type Netblock struct {
+	Address net.IP		`json:"address"`
+	Cidr int		`json:"cidr"`
+	Netmask net.IP		`json:"netmask"`
+	Gateway net.IP		`json:"gateway"`
+	Address_family int	`json:"address_family"`
```
Comment (crawford) [inline datasource/metadata/packet/metadata.go]:
```go
AddressFamily int	`json:"address_family"`
```

File: datasource/metadata/packet/metadata.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+type NetworkData struct {
+	Interfaces []Nic	`json:"interfaces"`
+	Netblocks []Netblock	`json:"addresses"`
+	Dns	[]net.IP	`json:"dns"`
+}
+
+// Metadata that will be pulled from the https://metadata.packet.net/metadata only. We have the opportunity to add more later.
+type Metadata struct {
+	Hostname string	  		`json:"hostname"`
+	Ssh_keys []string		`json:"ssh_keys"`
```
Comment (crawford) [inline datasource/metadata/packet/metadata.go]:
```go
SSHKeys []string		`json:"ssh_keys"`
```

File: datasource/metadata/packet/metadata.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+type Nic struct {
+	Name string	`json:"name"`
+	Mac string	`json:"mac"`
+}
+
+type NetworkData struct {
+	Interfaces []Nic	`json:"interfaces"`
+	Netblocks []Netblock	`json:"addresses"`
+	Dns	[]net.IP	`json:"dns"`
```
Comment (crawford) [inline datasource/metadata/packet/metadata.go]:
```go
DNS	[]net.IP	`json:"dns"`
```

**Comment by stresler** [comment]
Going to want this to go with this https://github.com/coreos/coreos-overlay/pull/1282

File: network/packet.go [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+		})
+		if netblock.Public == false {
+			routes = append(routes, route{
+				destination: net.IPNet{
+					IP:   net.IPv4(10, 0, 0, 0),
+					Mask: net.IPv4Mask(255, 0, 0, 0),
+				},
+				gateway: netblock.Gateway,
+			})
+		}
```
Comment (crawford) [inline network/packet.go]: If else

**Comment by stresler** [comment]
Sure.  Incoming update.

File: datasource/metadata/packet/metadata.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+	userdataUrl    = "userdata"
+	metadataPath   = "metadata"
+)
+
+type Netblock struct {
+	Address        net.IP `json:"address"`
+	Cidr           int    `json:"cidr"`
+	Netmask        net.IP `json:"netmask"`
+	Gateway        net.IP `json:"gateway"`
+	Address_Family int    `json:"address_family"`
```
Comment (crawford) [inline datasource/metadata/packet/metadata.go]: AddressFamily int    `json:"address_family"`

File: datasource/metadata/packet/metadata.go [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+		return
+	}
+
+	if len(m.NetworkData.Netblocks) > 0 {
+		for _, Netblock := range m.NetworkData.Netblocks {
+			if Netblock.Address_Family == 4 {
+				if Netblock.Public == true {
+					metadata.PublicIPv4 = Netblock.Address
+				}
+				if Netblock.Public == false {
```
Comment (crawford) [inline datasource/metadata/packet/metadata.go]: Else

File: datasource/metadata/packet/metadata.go [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+			}
+		}
+	}
+	metadata.Hostname = m.Hostname
+	metadata.SSHPublicKeys = map[string]string{}
+	for i, key := range m.SSHKeys {
+		metadata.SSHPublicKeys[strconv.Itoa(i)] = key
+	}
+
+	n, err := json.Marshal(m.NetworkData)
```
Comment (crawford) [inline datasource/metadata/packet/metadata.go]:
```go
metadata.NetworkConfig, err = json.Marshal(m.NetworkData)
```

This way you don't need the temporary and you don't shadow "err".

File: datasource/metadata/packet/metadata.go [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+
+	metadata.NetworkConfig = n
+
+	return metadata, err
```
Comment (crawford) [inline datasource/metadata/packet/metadata.go]:
```go
return
```

File: network/packet.go [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package network
+
+import (
+	"encoding/json"
+	"github.com/coreos/coreos-cloudinit/datasource/metadata/packet"
```
Comment (crawford) [inline network/packet.go]: Please move this below "net" with a newline separating them.

File: datasource/metadata/packet/metadata.go [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package packet
+
+import (
+	"encoding/json"
+	"github.com/coreos/coreos-cloudinit/datasource"
```
Comment (crawford) [inline datasource/metadata/packet/metadata.go]: Please move these two below "strconv" with a newline separating them.

**Comment by stresler** [comment]
I think that covers all the notes. Thanks for the help!

File: network/packet.go [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+		}
+	}
+
+	// Setup the variables we need to create our interfaces.
+	var b logicalInterface
+	var slaves []string
+	var children []networkInterface
+	var bmac net.HardwareAddr
+	var opts map[string]string
+	opts = make(map[string]string)
```
Comment (crawford) [inline network/packet.go]:
```go
opts := map[string]string {
    "Mode": "802.3ad",
    "LACPTransmitRate": "fast",
    ...
}
```

File: network/packet.go [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+	opts["MIIMonitorSec"] = ".2"
+	opts["UpDelaySec"] = ".2"
+	opts["DownDelaySec"] = ".2"
+
+	// Finish variables
+	for _, iface := range netdata.Interfaces {
+		if iface.Name != "chassis0" && iface.Name != "ipmi0" {
+			slaves = append(slaves, iface.Name)
+			if iface.Name == "enp1s0f0" {
+				bmac, _ = net.ParseMAC(iface.Mac)
```
Comment (crawford) [inline network/packet.go]: Just assign directly to `b.hwaddr` and drop the bmac variable.

**Comment by crawford** [comment]
@stresler I submitted a PR against your branch which has a few other changes. I figured it would be easier than going back and forth.

**Comment by stresler** [comment]
Yeah. Good changes and I see why you'd do it that way instead.  Once the tests pass I can sqaush. Thanks.

**Comment by crawford** [comment]
Can you also change your commit message to "datasource: add packethost metadata" once it's squashed?

**Comment by crawford** [comment]
@stresler one last lingering "else if". Otherwise, LGTM.

File: network/packet.go (hunk 4 above)
Comment (crawford) [inline network/packet.go]: Bump

**Comment by stresler** [comment]
Looks like there were 2 lingering ones.

**Comment by crawford** [comment]
LGTM!

**Pull request merged**